### PR TITLE
Basic Authentication

### PR DESCRIPTION
**Pull request by cliveglitch**
# Authentication Pull Request
I think I got a lot of progress with the authentication, and I wanted to receive some feedback about what I have.

## Features:
- Login/Signup modal.
- You can Login or Signup with credentials.
- You can Login or Signup with Discord, Google and GitHub providers if you fill the clientId and secret in the .env file for each provider.
- If you create an account by any means, like first login in with a provider or signing up, next auth will create a default player with the name of the user.
- The front end should tell you why you couldn't login or signup if you made a mistake, like wrong user or password, validations, if there is already a user with that email, etc.
- "your-games" will always default to the logged in user's first player.
- You can create games with any new user.
- Protected Page wrapper that will redirect you to the login page if you aren't logged in, example the your-games page.
- Passwords are saved as salted hashes with bcrypt.

## Does not have:
- Sessions.
- Forgot password functionality.
- Role based authorization.

image
image
image
image
image

### CODE REVIEW

File: next.config.js [hunk 1]
```diff
@@ -4,3 +4,8 @@ module.exports = {
   eslint: { ignoreDuringBuilds: !!process.env.CI },
   reactStrictMode: false,
 };
+module.exports = {
+  env: {
+    NEXTAUTH_URL: process.env.NEXTAUTH_URL,
```
Comment (FunctionDJ) [inline next.config.js:9]: why only this one? shouldn't next auth read the environment variables regardless?

File: prisma/schema.prisma [hunk 2]
```diff
@@ -39,12 +39,37 @@ enum Role {
...
+  email         String?   @unique
+  emailVerified DateTime?
+  image         String?
+  createdAt     DateTime  @default(now())
+  updatedAt     DateTime  @updatedAt
+  identities    Player[]
+  accounts      Account[]
+}
+
+model Account {
```
Comment (FunctionDJ) [inline prisma/schema.prisma:56]: if a user uses both discord and email to login, they will have 2 accounts but 1 user on the DB, right?
how would you match them up?

File: prisma/schema.prisma [hunk 3]
```diff
@@ -39,12 +39,37 @@ enum Role {
...
+  access_token      String? @db.Text
+  expires_at        Int?
+  token_type        String?
+  scope             String?
+  id_token          String? @db.Text
+  session_state     String?
+
+  user User @relation(fields: [userId], references: [id], onDelete: Cascade)
+
+  @@unique([provider, providerAccountId])
```
Comment (FunctionDJ) [inline prisma/schema.prisma:72]: intuitively, i'd say the unique should be on `[provider, userId]`, no? and then do we even need the `id` anymore when these entrys can and need to be idenfitied by those two properties/columns?

File: prisma/schema.prisma [hunk 4]
```diff
@@ -88,7 +113,7 @@ model Player {
   secret      Boolean       @default(false) // secret = true means that parent user is hidden on profile
   ChatMessage ChatMessage[]
 
-  User   User   @relation(fields: [userId], references: [id])
+  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade)
```
Comment (FunctionDJ) [inline prisma/schema.prisma:116]: please explain. to me this looks dangerous and would attempt to delete the user when the player gets deleted. or is it the other way around?

File: src/frontend/context/players.tsx [hunk 5]
```diff
@@ -1,41 +1,67 @@
...
       {children}
     </playersContext.Provider>
   );
 };
 
 export const usePlayers = () => {
-  // TODO make sure development_player0 is set by default - currently not.
-  const ownedPlayers = useContext(playersContext);
+  const user = useContext(playersContext);
+  const ownedPlayers = user?.ownedPlayers;
```
Comment (FunctionDJ) [inline src/frontend/context/players.tsx:49]: have you implemented a "select default player" mechanic like the removed comment said on L25?

File: src/pages/api/WarsWorldAdapter.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
+// Based on Prisma Adapter
+
+import type { PrismaClient } from "@prisma/client";
+import type { Adapter, AdapterAccount } from "@auth/core/adapters";
+import { AdapterUser } from "next-auth/adapters";
+import { Awaitable } from "next-auth";
+
+export default function WarsWorldAdapter(p: PrismaClient): Adapter {
```
Comment (FunctionDJ) [inline src/pages/api/WarsWorldAdapter.ts:8]: can you please explain how this works and why it needs an adapter? (basically, pls explain next-auth lol sorry)

File: src/pages/api/authOptions.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            email: dbUser.email,
+          };
+          return dbUserWithoutPassword as User;
+        }
+
+        return null;
+      },
+    }),
+    GithubProvider({
+      clientId: process.env.GITHUB_CLIENT_ID as string,
```
Comment (FunctionDJ) [inline src/pages/api/authOptions.ts:67]: how does this handle when you don't specify the necessary environment variables? will there be errors? does the login option still show up?

File: src/pages/api/jwt.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+// This is an example of how to read a JSON Web Token from an API route
+import { getToken } from "next-auth/jwt";
+
+import type { NextApiRequest, NextApiResponse } from "next";
+
+export default async function handler(
```
Comment (FunctionDJ) [inline src/pages/api/jwt.ts:6]: i'd prefer not using `default` to incentivise using the actual name of this function, which probably should be something other than `handler`. or if it's truly just an example, the file name should be different.

File: src/pages/api/jwt.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  // If you don't have the NEXTAUTH_SECRET environment variable set,
```
Comment (FunctionDJ) [inline src/pages/api/jwt.ts:10]: when would this case come up?

File: src/server/hashPassword.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { genSalt, hash } from "bcrypt";
+
+const SALT_ROUNDS = 12;
+
+export default async function hashPassword(password: string) {
```
Comment (FunctionDJ) [inline src/server/hashPassword.ts:5]: no `default` pls

File: src/server/routers/match/create.ts [hunk 11]
```diff
@@ -6,18 +6,21 @@ import { prisma } from "server/prisma/prisma-client";
...
 import { PlayerInMatch } from "shared/types/server-match-state";
+import { withPlayerNameSchema } from "server/trpc/middleware/player";
 
 export const createMatchProcedure = playerBaseProcedure
   .input(
     withMapIdSchema.extend({
       selectedCO: coSchema,
     })
   )
+  .input(withPlayerNameSchema)
```
Comment (FunctionDJ) [inline src/server/routers/match/create.ts:17]: these schemas appear to be used only once, so it's not useful to extract them into reusable variables yet. also you're extending one schema and then using a second `.input` for another one, kind mix-and-matching. i think the mapId, selectedCO and playerName can just be a part of a single inline schema in this case.

File: src/server/schemas/auth.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import { z } from "zod";
+
+// TODO: better password zod validation
+
+export const loginSchema = z.object({
+  username: z.string().min(1),
+  password: z.string().min(4).max(20),
```
Comment (FunctionDJ) [inline src/server/schemas/auth.ts:7]: we *can* add better password validation here, but i think it's simpler to just do that in the procedure layer. also i think the range should be something more like `8-40`.

File: src/server/schemas/auth.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+});
+
+export const signUpSchema = loginSchema.extend({
+  email: z.string().email(),
+  confirmPassword: z.string().min(4).max(20),
```
Comment (FunctionDJ) [inline src/server/schemas/auth.ts:12]: `confirm`? on the backend, this would just be `password` to us. the confirm part is only frontend.

File: src/server/schemas/auth.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+});
+
+export type ILogin = z.infer<typeof loginSchema>;
```
Comment (FunctionDJ) [inline src/server/schemas/auth.ts:15]: unless other types have been prefixed with `I`, it'd be consistent to remove them.

File: src/server/trpc/middleware/player.ts [hunk 15]
```diff
@@ -68,10 +71,7 @@ export const playerMiddleware = t.middleware(async ({ ctx, next, input }) => {
 
 export const playerWithoutCurrentMiddleware = t.middleware(
   async ({ ctx, next }) => {
-    const ownedPlayers =
-      process.env.NODE_ENV === "development"
-        ? await getDevelopmentModeUserPlayers()
-        : await getProductionModeUserPlayers(ctx.session);
+    const ownedPlayers = await getProductionModeUserPlayers(ctx.session);
```
Comment (FunctionDJ) [inline src/server/trpc/middleware/player.ts:74]: so there's no getDevModePlayers anymore? if the function handles this internally, the name should be updated. otherwise, dev mode players are missing now.

File: src/server/trpc/trpc-init.ts [hunk 16]
```diff
@@ -1,10 +1,20 @@
...
-  errorFormatter({ shape }) {
-    return shape;
+  errorFormatter({ shape, error }) {
+    return {
+      ...shape,
+      data: {
+        ...shape.data,
+        zodError:
+          error.code === "BAD_REQUEST" && error.cause instanceof ZodError
+            ? error.cause.flatten()
```
Comment (FunctionDJ) [inline src/server/trpc/trpc-init.ts:15]: i know this is from the trpc docs but some de-nesting could improve readability here.

**Review by FunctionDJ — CHANGES_REQUESTED** [review]

File: src/pages/api/authOptions.ts (hunk 7 above)
Comment (cliveglitch) [reply src/pages/api/authOptions.ts:67]: You simply cannot log in with those providers. You don't need to specify them all, if you want to try to log in with GitHub, just fill GitHub's env variables, and if you want to log in with email and password you don't need to fill any.

File: prisma/schema.prisma (hunk 2 above)
Comment (cliveglitch) [reply prisma/schema.prisma:56]: Next Auth looks for the tables called User and Account to login and signup a user and that's why the naming, Accounts are all the providers the user have tied, like Discord, GitHub and Google. 

For now there is no functionality to bind two emails to the same user at the moment, that means if you try to sign up with email/password and then you try to login for the first time with discord it will throw an error. Also email signup is not made into an account, it just fills the password so the user can login with a password.

In short, you sign up as one user, that user can have cero to many accounts (like discord), and that user can have 1 to many players, players being the entities that can play the game.

File: prisma/schema.prisma (hunk 3 above)
Comment (cliveglitch) [reply prisma/schema.prisma:72]: I've took this schema from NextAuth's documentation: https://next-auth.js.org/v3/adapters/prisma
And I think it is what NextAuth expects for this to work.

File: prisma/schema.prisma (hunk 4 above)
Comment (cliveglitch) [reply prisma/schema.prisma:116]: Deletes the player when the user is deleted, I just put that there to make it more convenient to delete users and test the functionality, is it fine?

File: src/frontend/context/players.tsx (hunk 5 above)
Comment (cliveglitch) [reply src/frontend/context/players.tsx:49]: Yes, even if you create a new User, when you enter the page the first player will be selected.

File: next.config.js (hunk 1 above)
Comment (cliveglitch) [reply next.config.js:9]: I've continued Matthew's work and that is something he left there, it stills works fine without it but I cannot say why it's needed there.

File: src/pages/api/jwt.ts (hunk 8 above)
Comment (cliveglitch) [reply src/pages/api/jwt.ts:6]: I understand, didn't had that on mind while trying to create my functions, I just added the default to the main function of the file, but I follow you.

File: src/server/trpc/middleware/player.ts (hunk 15 above)
Comment (cliveglitch) [reply src/server/trpc/middleware/player.ts:74]: Ok, change the name, I used only getProductionModeUserPlayers so I can see the players of any new user that doesn't have that prefix.

File: src/pages/api/jwt.ts (hunk 9 above)
Comment (cliveglitch) [reply src/pages/api/jwt.ts:10]: Matthew's file, I think it'd come up with sessions, but for now I plan to leave with sessions in a future PR, so I'm inclined to delete this file.

File: src/server/routers/match/create.ts (hunk 11 above)
Comment (cliveglitch) [reply src/server/routers/match/create.ts:17]: Doesn't seem like the name input is needed in the first place, I think I left it there when trying to show the player's name instead of their Id on those match cards but kinda forgot.

File: src/server/schemas/auth.ts (hunk 13 above)
Comment (cliveglitch) [reply src/server/schemas/auth.ts]: hehe I was kinda playing with zod and what I can do, the latest commit should move the confirm check to the front.

File: src/server/trpc/trpc-init.ts (hunk 16 above)
Comment (cliveglitch) [reply src/server/trpc/trpc-init.ts:15]: mmm... how would you de-nest it?

File: src/server/hashPassword.ts (hunk 10 above)
Comment (cliveglitch) [reply src/server/hashPassword.ts]: Done

File: src/server/schemas/auth.ts (hunk 14 above)
Comment (cliveglitch) [reply src/server/schemas/auth.ts]: Done

File: next.config.js (hunk 1 above)
Comment (FunctionDJ) [reply next.config.js]: then it should be removed.

File: prisma/schema.prisma (hunk 4 above)
Comment (FunctionDJ) [reply prisma/schema.prisma:116]: it's probably fine, but if a user gets deleted, we need to do cleanups regardless.
players can have clan relationships, which need to be resolved. the "ownedClans" one can't be cascade-deleted i think.
i'm not sure if we should use cascade deletion or just do everything manually because we can't fully "automate" it anyway.

File: src/frontend/components/auth/SignupForm.tsx [hunk 17]
```diff
@@ -22,6 +22,8 @@ export default function SignupForm({
     confirmPassword: "",
   });
 
+  const [doPasswordsMatch, setDoPasswordsMatch] = useState(true);
```
Comment (FunctionDJ) [inline src/frontend/components/auth/SignupForm.tsx:25]: i don't think state is needed. just do `passwordsMatch = signupData.password === signupData.confirmPassword`

File: src/pages/api/authOptions.ts (hunk 7 above)
Comment (FunctionDJ) [reply src/pages/api/authOptions.ts:73]: i've tested it and the buttons show up and i think clicking them will start the oauth flow which will immediately just redirect to the application (like a refresh) when these environment variables are are missing/invalid.

i think it would be a good idea if we made this more explicit in the GUI (e.g. for new developers) where the buttons are greyed out or just removed if the credentials are not set.

the latter might be easy to achieve by padding the providers conditionally like we've done in the past at points.
what do you think?

File: src/server/trpc/trpc-init.ts (hunk 16 above)
Comment (FunctionDJ) [reply src/server/trpc/trpc-init.ts:15]: i think pulling the value of `zodError` into a variable makes sense.

File: src/pages/api/WarsWorldAdapter.ts (hunk 6 above)
Comment (FunctionDJ) [reply src/pages/api/WarsWorldAdapter.ts:8]: so when someone uses email/password to sign up, they don't get a default player, right?
can we do something about that?

Comment (cliveglitch) [reply src/pages/api/WarsWorldAdapter.ts:8]: They do, and I already have that done, you just need to add that logic separately.
NextAuth is not so keen with credentials login with password, so they design it to take more effort to set it up purposefully.

File: src/pages/api/authOptions.ts (hunk 7 above)
Comment (cliveglitch) [reply src/pages/api/authOptions.ts:73]: sounds like a good idea, I'm gonna work on it, later I'll tell you when it's done.

File: prisma/schema.prisma (hunk 4 above)
Comment (cliveglitch) [reply prisma/schema.prisma:116]: would need some time to investigate this, for now I got this for testing purposes so we could change if it has no real use.

Comment (cliveglitch) [reply prisma/schema.prisma]: I opted to remove the the onCascade deletion for now, if I can find a way to automatically delete a user completely then we can discuss that later.

File: src/pages/api/authOptions.ts (hunk 7 above)
Comment (cliveglitch) [reply src/pages/api/authOptions.ts:96]: Done, it now greys out the buttons when you don't have the necessary env variables for that provider. 
In this example I don't have the env for Google's provider.
image

File: src/server/trpc/trpc-init.ts (hunk 16 above)
Comment (cliveglitch) [reply src/server/trpc/trpc-init.ts]: Done

File: src/frontend/components/auth/SignupForm.tsx (hunk 17 above)
Comment (cliveglitch) [reply src/frontend/components/auth/SignupForm.tsx]: Done

File: src/frontend/components/modals/LoginSignupModal.tsx [hunk 18]
```diff
@@ -17,13 +23,29 @@ interface Props {
 }
 
 export default function LoginSignupModal({ isOpen, setIsOpen, width }: Props) {
+  const posibleProviders = ["github", "discord", "google"];
+  const [currentProviders, setCurrentProviders] = useState<Record<
+    LiteralUnion<BuiltInProviderType, string>,
```
Comment (FunctionDJ) [inline src/frontend/components/modals/LoginSignupModal.tsx:28]: maybe this type expression can be simplified with something like
```ts
Awaited<ReturnType<typeof getProviders>>
```
but i'll leave this up to you if it's better or not (if it works)

File: src/frontend/components/modals/LoginSignupModal.tsx [hunk 19]
```diff
@@ -47,6 +69,21 @@ export default function LoginSignupModal({ isOpen, setIsOpen, width }: Props) {
 
   const onClose = async () => await setIsOpen(false);
 
+  const renderProviders = () =>
```
Comment (FunctionDJ) [inline src/frontend/components/modals/LoginSignupModal.tsx:72]: inline components should be avoided because of re-renders (and it would also break hooks/state). just make `currentProviders` a prop. you don't need to turn `possibleProviders` into a prop since it can live outside of components anyway.

File: src/pages/api/authOptions.ts [hunk 20]
```diff
@@ -13,60 +13,101 @@ import { Adapter } from "next-auth/adapters";
...
-          };
-          return dbUserWithoutPassword as User;
-        }
+    },
+    async authorize(credentials) {
+      if (!credentials || !credentials.name || !credentials.password) {
+        return null;
+      }
+
+      const creds = await loginSchema.parseAsync({
```
Comment (FunctionDJ) [inline src/pages/api/authOptions.ts:62]: this code could be simplified by making the submitted data fit to the loginSchema or vice versa.
then you could do `const { username, password } = loginSchema.safeParse(credentials)`
([don't use `async` unless the schema uses async refiners etc.](https://github.com/colinhacks/zod#parseasync))
(also i think [.strict()](https://github.com/colinhacks/zod#strict) would be a good idea here to ban needless properties which could indicate issues, up to you whether to apply it here when used or to put it on `loginSchema`'s definition directly)

File: src/pages/api/authOptions.ts [hunk 21]
```diff
@@ -13,60 +13,101 @@ import { Adapter } from "next-auth/adapters";
...
-      },
-    }),
+      }
+
+      const doPasswordsMatch = await compare(
+        credentials.password,
+        dbUser.password
+      );
+
+      if (doPasswordsMatch) {
```
Comment (FunctionDJ) [inline src/pages/api/authOptions.ts:80]: i think code makes more sense with early-returns in "failure cases", related to the "fail fast" principle.
in practise, that'd mean `if (!doPasswordsMatch) { return null; } ...`

File: src/pages/api/authOptions.ts [hunk 22]
```diff
@@ -84,12 +125,14 @@ export const authOptions: NextAuthOptions = {
...
-      clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
-    }),
-  ],
+    })
+  );
+
+export const authOptions: NextAuthOptions = {
+  secret: process.env.NEXTAUTH_SECRET,
+  adapter: adapter,
+  debug: true,
```
Comment (FunctionDJ) [inline src/pages/api/authOptions.ts:134]: maybe something like `debug: process.env.NODE_ENV === "dev"` ?
idk which env vars are currently available to check for development.
maybe just `console.log(process.env)` and see if something's in there.

**Review by FunctionDJ — CHANGES_REQUESTED** [review]

File: src/frontend/components/modals/LoginSignupModal.tsx [hunk 23]
```diff
@@ -17,13 +23,29 @@ interface Props {
...
   const searchParams = useSearchParams();
   const router = useRouter();
 
   const [didSignUp, setDidSignUp] = useState(false);
+
   const isSignupForm = searchParams.has("SignUpForm");
   const callbackUrl = searchParams.get("callbackUrl");
 
+  useEffect(() => {
+    const setProviders = async () => {
```
[the remark this answers is not in the document]
Comment (cliveglitch) [reply src/frontend/components/modals/LoginSignupModal.tsx]: Much simpler.

File: src/frontend/components/modals/LoginSignupModal.tsx (hunk 18 above)
Comment (cliveglitch) [reply src/frontend/components/modals/LoginSignupModal.tsx]: It works, so it's worth it to change it.

File: src/pages/api/authOptions.ts (hunk 22 above)
Comment (cliveglitch) [reply src/pages/api/authOptions.ts]: Done.

File: src/pages/api/authOptions.ts (hunk 20 above)
Comment (cliveglitch) [reply src/pages/api/authOptions.ts]: Decided to leave out the .strict(), but everything else is done.

File: src/frontend/components/modals/LoginSignupModal.tsx (hunk 19 above)
Comment (cliveglitch) [reply src/frontend/components/modals/LoginSignupModal.tsx]: I've just put back the map into the jsx

File: src/frontend/components/modals/LoginSignupModal.tsx [hunk 24]
```diff
@@ -17,13 +23,29 @@ interface Props {
 }
 
 export default function LoginSignupModal({ isOpen, setIsOpen, width }: Props) {
+  const posibleProviders = ["github", "discord", "google"];
```
[the remark this answers is not in the document]
Comment (cliveglitch) [reply src/frontend/components/modals/LoginSignupModal.tsx]: Done.

File: src/frontend/components/auth/SignupForm.tsx [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+    user: "",
+    email: "",
+    password: "",
+    confirmPassword: "",
+  });
+
+  const doPasswordsMatch =
+    signupData.password === signupData.confirmPassword ||
+    signupData.password === "" ||
+    signupData.confirmPassword === "";
```
[the remark this answers is not in the document]
Comment (cliveglitch) [reply src/frontend/components/auth/SignupForm.tsx]: I think it's a misunderstanding, I've changed the code a bit to make it more obvious, but the purpose of that conditional is when to `show` an error, the logic is, when the passwords do not match and there is at least a character on both the inputs then it will show the error.
It's to not bother the user with a "Passwords do not match message when they are typing the first password".

File: src/frontend/components/auth/SignupForm.tsx [hunk 26]
```diff
@@ -38,12 +43,15 @@ export default function SignupForm({
   const onSubmitSignupForm = async (event: FormEvent) => {
     event.preventDefault();
 
+    if (signupData.confirmPassword !== signupData.password) {
```
[the remark this answers is not in the document]
Comment (cliveglitch) [reply src/frontend/components/auth/SignupForm.tsx]: You can submit when passwords do not match, there is no code that stops the user not to submit until that point.
The code on the review above was to show the error.

File: src/frontend/components/auth/SignupForm.tsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (FunctionDJ) [reply src/frontend/components/auth/SignupForm.tsx]: Ah, i get it now. Please change the variable to something like
```ts
const showPasswordsDoNotMatch = signupData.password !== "" && signupData.confirmPassword !== "" && signupData.password !== signupData.confirmPassword
```
This makes the logic and the intent behind this code much easier to understand, because the variable isn't exclusively about whether the passwords match, but also if it makes sense to show a "passwords don't match" hint.

File: src/frontend/components/auth/SignupForm.tsx (hunk 26 above)
[the remark this answers is not in the document]
Comment (FunctionDJ) [reply src/frontend/components/auth/SignupForm.tsx]: Why not disable the submit button when passwords don't match instead?

File: src/frontend/components/modals/LoginSignupModal.tsx (hunk 19 above)
Comment (FunctionDJ) [reply src/frontend/components/modals/LoginSignupModal.tsx]: yes, that's fine.

File: src/pages/api/authOptions.ts [hunk 27]
```diff
@@ -55,37 +55,37 @@ const providers: Provider[] = [
...
-        return dbUserWithoutPassword as User;
+      if (!doPasswordsMatch) {
+        return null;
       }
 
-      return null;
+      return {
+        name: dbUser.name,
+        email: dbUser.email,
+      } as User;
```
Comment (FunctionDJ) [inline src/pages/api/authOptions.ts:88]: What's up with this type override?

File: src/frontend/components/auth/SignupForm.tsx (hunk 26 above)
[the remark this answers is not in the document]
Comment (cliveglitch) [reply src/frontend/components/auth/SignupForm.tsx]: Kinda does the same I suppose, you just get rejected after you click it. I've change it no problem, and also added so you cannot press the button if you have empty fields because it doesn't make sense to disable it only with the passwords.

File: src/frontend/components/auth/SignupForm.tsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (cliveglitch) [reply src/frontend/components/auth/SignupForm.tsx]: Done.

File: src/pages/api/authOptions.ts (hunk 27 above)
Comment (cliveglitch) [reply src/pages/api/authOptions.ts]: You get this error otherwise:
image

You can add the id to the object (I did that on the last commit), and it works the same.

**Review by FunctionDJ — APPROVED** [review]
